### ISSUE TITLE
Question About Pthreadpool (Seems Bug)

### ISSUE THREAD

**Issue by vippeterhou** [google/XNNPACK#883]
Hi, 

I build a CNN model which has three convolution layers.

I use pthreadpool as follows (the platform is ARM architecture):

```cpp
auto num_cores = 1;
auto threads = pthreadpool_create(num_cores);

// init
xnn_initialize(nullptr /* allocator */);

// create
xnn_status status = xnn_create_convolution2d_nchw_f32(...);

// setup
xnn_setup_convolution2d_nchw_f32(..., threads /* thread poll */)

// inference
xnn_run_operator(..., threads /* thread pool */);
```

When the `num_cores` is 1 (if I pass `threads` as `nullptr`, it uses 1 core by default.). The result is correct and everything is fine. 

However, if I set the `num_cores` to value larger than 1 (whatever 2 or 6, or others), the result is wrong. 

There are two things I want to highlight in this error:

1. In my case, the result `res` has dim: 1, 16, 240, 240. The `res[0][0][:][:]` is correct, and others are wrong (most of them are zeros).
2. The first and second convolutions' outputs are correct. The incorrectness occurs from the third convolution. 

According to my observations, I have three questions:
1. I want to know whether I missed something or used pthreadpool in the wrong way?
2. Did you have any test/benchmark about multi-cpucores usage?
3. Does it need to synchronize when using multi-cpucores?

Thanks a lot!

**Comment by Maratyszcza** [comment]
From your description it does look like a bug. Please check that you pass the same `pthreadpool_t` object to both `xnn_setup_convolution2d_nchw_f32` and `xnn_run_operator`, and share the following details:
1. Parameters of the first convolution which produce wrong results.
2. The platform you tested on (x86 or ARM, and model name of the processor).

> According to my observations, I have three questions:
> 
> I want to know whether I missed something or used pthreadpool in the wrong way?
> Did you have any test/benchmark about multi-cpucores usage?
> Does it need to synchronize when using multi-cpucores?
1. I don't see any obviously wrong.
2. We do have multi-threaded tests, albeit most tests run as single-threaded.
3. When `xnn_run_operator` returns, all threads should have the same synchronized view of the memory.

**Comment by vippeterhou** [comment]
Thanks for your reply!

I think I have found out where the bug occurs.

I tested several different CNN models and I found the error always occurs in the first 1x1 convolution. 

Then I built a one-layer model that only has a 1x1 convolution.

As expected, when the `num_cores` is larger than 1, the result will be wrong.

As for the parameters of this 1x1 convolution, both input and output have the dimension of `1, 16, 80, 160`, specifically:
```bash
padding: 0, 0, 0, 0
kernel h, kernel w: 1, 1
stride h, stride w: 1, 1
input channel: 16
output channel: 16
group: 1
dilation h, dilation w: 1, 1
batch size: 1
```

My platform is ARM architecture (NVIDIA Xavier). 

Hope it helps.

**Comment by Maratyszcza** [comment]
I just noticed that you use NCHW convolution. NCHW operators in XNNPACK are intended for sparse inference (see [Fast Sparse ConvNets paper](https://arxiv.org/abs/1911.09723)). They are not efficient for dense models.

**Comment by vippeterhou** [comment]
Yes, I understand I'm using NCHW operators, and I indeed will extend to sparse inference in the future (if the dense result is correct).

I'm currently working on dense models for evaluating the correctness of XNNPACK. However, I think it should have the correct result, albeit it is not efficient at this stage.

I also expertise in the pipeline of sparse convolution. In my case, the input must be NHWC. The first convolution should use the `XNN_FLAG_INPUT_NHWC` flag. And the second is depthwise-convolution. The problem is from the third 1x1 convolution as I mentioned before. 

So I think it may have something wrong in 1x1 convolution when handling the multi-thread. Could you please find out the problem? Thanks a lot!

**Comment by vippeterhou** [comment]
Hi, 

Do you have any updates? Don't hesitate to contact me if you have any questions when reproducing the error. Thanks!

**Comment by Maratyszcza** [comment]
I reproduced the error, but it will take a while to fix. ETA 1 month.

**Comment by Maratyszcza** [comment]
Should be fixed in e8bfcc8da8c5644b620c33a3b034f66d07e033f6
